File: arduinobot_ws/src/arduinobot_firmware/src/simple_serial_transmitter.cpp
/**
 * @file simple_serial_transmitter.cpp
 * @brief ROS 2 Node that receives ROS messages and writes them to a serial port.
 *
 * This node subscribes to a String topic and relays any received data directly
 * over serial to an attached Arduino.
 */

#include <rclcpp/rclcpp.hpp>
#include <std_msgs/msg/string.hpp>
#include <libserial/SerialPort.h>

using std::placeholders::_1;

/**
 * @class SimpleSerialTransmitter
 * @brief Node that subscribes to 'serial_transmitter' and forwards incoming messages to a serial interface.
 */
class SimpleSerialTransmitter : public rclcpp::Node
{
public:
  /**
   * @brief Construct a new Simple Serial Transmitter object.
   *
   * Declares target port parameters, subscribes to the transmitter topic, and open
   * the serial connection.
   */
  SimpleSerialTransmitter() : Node("simple_serial_transmitter")
  {
    declare_parameter<std::string>("port", "/dev/ttyACM0");
    std::string port_ = get_parameter("port").as_string();
    
    sub_ = create_subscription<std_msgs::msg::String>(
        "serial_transmitter", 10, std::bind(&SimpleSerialTransmitter::msgCallback, this, _1));
    
    arduino_.Open(port_);  
    arduino_.SetBaudRate(LibSerial::BaudRate::BAUD_115200);    
  }

  /**
   * @brief Destroy the Simple Serial Transmitter object.
   *
   * Ensures that the serial communication port is closed.
   */
  ~SimpleSerialTransmitter()
  {
    arduino_.Close();
  }

  /**
   * @brief Callback function triggered when a new String message is received.
   *
   * Writes the message payload bytes onto the active serial line.
   *
   * @param msg The incoming String message.
   */
  void msgCallback(const std_msgs::msg::String &msg)
  {
    RCLCPP_INFO_STREAM(get_logger(), "New message recieved, publishing on serial port: " << msg.data);
    arduino_.Write(msg.data);
  }

private:
  rclcpp::Subscription<std_msgs::msg::String>::SharedPtr sub_;
  LibSerial::SerialPort arduino_;
};


int main(int argc, char * argv[])
{
  rclcpp::init(argc, argv);
  rclcpp::spin(std::make_shared<SimpleSerialTransmitter>());
  rclcpp::shutdown();
  return 0;
}

File: arduinobot_ws/src/arduinobot_firmware/arduinobot_firmware/simple_serial_transmitter.py
#!/usr/bin/env python3
"""
Module providing a ROS 2 node to transmit serial messages to Arduino.
"""
import rclpy
from rclpy.node import Node
from std_msgs.msg import String
import serial


class SimpleSerialTransmitter(Node):
    """
    ROS 2 Node that subscribes to a topic and forwards incoming string messages to a serial port.
    """

    def __init__(self):
        """
        Initialize the SimpleSerialTransmitter node.

        Declares 'port' and 'baud_rate' parameters, opens the serial connection,
        and sets up a subscription to the 'serial_transmitter' topic.
        """
        super().__init__("simple_serial_transmitter")
        
        self.declare_parameter("port", "/dev/ttyACM0")
        self.declare_parameter("baud_rate", 115200)

        self.port_ = self.get_parameter("port").value
        self.baud_rate_ = self.get_parameter("baud_rate").value
        
        self.sub_ = self.create_subscription(String, "serial_transmitter", self.msgCallback, 10)
        self.arduino_ = serial.Serial(port=self.port_, baudrate=self.baud_rate_, timeout=0.1)

    def msgCallback(self, msg):
        """
        Callback triggered whenever a new String message is received.

        Encodes the message data as UTF-8 and writes it directly to the serial port.

        :param msg: The incoming String message.
        """
        self.get_logger().info("I heard: %s" % msg.data)
        self.arduino_.write(msg.data.encode("utf-8"))


def main():
    rclpy.init()

    simple_serial_transmitter = SimpleSerialTransmitter()
    rclpy.spin(simple_serial_transmitter)
    
    simple_serial_transmitter.destroy_node()
    rclpy.shutdown()


if __name__ == '__main__':
    main()